File: 常用算法模板/二分查找/模板.cpp

typedef long long lld;

const lld M = 1e5+2;
class Solution {
public:

    bool checkFunc(vector<int>& time, int totalTrips, lld totalTime) {
        lld trips =0;
        for( auto t:time) {
            trips += totalTime/t;
        }

        return trips>=totalTrips;
    }

    long long minimumTime(vector<int>& time, int totalTrips) {
        lld best = 1e14+10;
        lld left=1, right= best;
        while( left<= right) {
                lld mid = (right+left)>>1;
                if ( checkFunc(time, totalTrips, mid)) { // 可行解，贪心查看有没有更好解, 区间往小移
                    best = mid;
                    right = mid-1;
                } else {
                    left = mid+1;
                }
        }

        return best;
    }
};

File: 常用算法模板/二分查找/模板.go
package main
import "fmt"


func abs(a int) int {
	if a<0 {return -a}
	return a
}
func checkFunc(houses []int, heaters []int, radius int) bool {
	i, j:=0,0
	for ;i<len(houses) && j<len(heaters); {
		if abs(houses[i] - heaters[j])<=radius{
			i++
		}else {j++}
	}
	return i==len(houses)
}

func findRadius(houses []int, heaters []int) int {
	// 排序
	sort.Slice(houses, func(i, j int) bool {
		return houses[i]<houses[j]
	})
	sort.Slice(heaters, func(i, j int) bool {
		return heaters[i]<heaters[j]
	})
	best := int(10e9)
	left, right := 0, int(10e9)
	for left<= right {
		mid := (right+left)>>1
		if checkFunc(houses, heaters, mid) { // 可行解，贪心查看有没有更好解, 区间往小移
			best, right = mid, mid-1
		} else {
			left = mid+1
		}
	}

	return best
}

/*
[1,2,3]
[2]
[2]
[2]
[1,2,3,4]
[1,4]
[1,5]
[2]
[4,3,2,1]
[1,4]
*/}